### PR TITLE
Copying tuples lazily while casting a monotonic heap

### LINKED ISSUE #50: Writing to tuples inside tuples in place when casted monotonically
Some versions of ray benchmark fails because of this. One way to fix it is to deep copy tuple values.

### PR DESCRIPTION
**Pull request by deyaaeldeen**
Fixes #50.

The old logic was copying tuples inside a monotonic heap once we do a monotonic cast but failed to do it recursively in the case of tuples of tuples. The new approach does the copying inside a tuple cast in a monotonic cast and does it only for the current layer of tuples so that it does not do excessive copying if the next cast does not require it. This approach fixes the bug and does not have to do a full copy if not necessary.

### CODE REVIEW

**Comment by deyaaeldeen** [comment]
Done.

**Comment by deyaaeldeen** [comment]
This is a bug fix not an optimization and I added a test case that you can run with master and this branch to see the difference. I did not support preventing the same tuple from being copied multiple times, I hope to do it in the future but this is an optimization for a corner case that does not show up much in practice. 

Now to describe the gist behind copying tuples, look at Fig 6 in the paper. cv' in CastRef1 is basically a pair of a casted value and a cast on it that should be written to the heap as is. This gets reduced later by the state reduction rules (notably HCAST and HDROP) until the heap has no more casted values. My implementation represents these casted values by a copy of the original tuple value on the heap. Then any cast on that cv writes the result to each component of the tuple right away without waiting until the full tuple cast finishes.

The bug was basically representing the top level casted values correctly by copying but not the subcomponent of it, so deeply nested tuples gets overwritten with the assumption that they were copied was false. Now, I pushed the copying logic from the point of a monotonic cast to the point of a tuple cast inside a monotonic cast, so all nested tuples will be casted in this case.

So to recap, there is no weird optimizations in my code, I am just trying to be clever with my runtime representation of casted values. If I just do what the paper says, I will add new forms of values to represent casted values and unnecessarily store cast information in them. Let me know if you have any more questions.

**Comment by akuhlens** [comment]
> So to recap, there is no weird optimizations in my code, I am just trying to be clever with my runtime representation of casted values. 

I would like comments to understand the purpose of the clever code. This will make it more maintainable and easier for others to perform reviews. 

> Now to describe the gist behind copying tuples, look at Fig 6 in the paper. cv' in CastRef1 is basically a pair of a casted value and a cast on it that should be written to the heap as is. This gets reduced later by the state reduction rules (notably HCAST and HDROP) until the heap has no more casted values. 

I am fine with requiring programmers to read the paper, instead of explaining the theory, but since the implementation diverges from the theory we may want to include a pointer to what theory we are using and the differences between our implementation and that theory.

> This is a bug fix not an optimization and I added a test case that you can run with master and this branch to see the difference.

I am not worried about understanding the bug fix, I am worried about understanding what the code does as a whole. With your current comments I can understand how these arguments are used, but to what end, why are we updating tuples and arrays in place?

> My implementation represents these casted values by a copy of the original tuple value on the heap. Then any cast on that cv writes the result to each component of the tuple right away without waiting until the full tuple cast finishes.

The reason I would like documentation for this code is because I would like to understand why this
is equivalent to the original, and what benefit we derive from it, it would also be nice to measure this benefit in some way. 

> If I just do what the paper says, I will add new forms of values to represent casted values and unnecessarily store cast information in them.

This may be more necessary than you think, but for reasons besides tuples (see test case below). At some point we had a conversation about what values required type information and which didn't. Based on my current understanding of that conversation, monotonic references are the only thing that would require type information. This and my current understanding of this code indicates that we may be able to simplify this code. Simultaneously, I am afraid that I don't understand everything that you have done, so I would like better documentation so that I can.

In the interim, here is a test case for this code which fails.

```racket
(print-int                                                                          
 (let ([r0 (mvector 2 (ann 666 Dyn))])                                              
   (let ([v : Dyn (tuple (ann 66 Dyn)                                               
                         (ann 6  Dyn)                                               
                         (ann r0 Dyn))])                                            
     (begin                                                                         
       (mvector-set! r0 0 v)                                   
       (let ([r1 : (MVect (Tuple Int Dyn (MVect (Tuple Int Int Dyn)))) r0])         
         (tuple-proj (mvector-ref r0 1) 0)))))) 
```

There are two MVect casts in this code. The first one implicitly erases that the second index is a dynamic value by writing the "t3" type information to the vector before the elements are casted,
then the second cast uses "t3" to cast the 2nd element, before the first casts has cast the 2nd
element.

**Comment by deyaaeldeen** [comment]
> I would like comments to understand the purpose of the clever code. This will make it more 
> maintainable and easier for others to perform reviews.

I already answered that in my previous comment literally in the next sentence.

> I am fine with requiring programmers to read the paper, instead of explaining the theory, but since 
> the implementation diverges from the theory we may want to include a pointer to what theory we are 
> using and the differences between our implementation and that theory.

Users of our language do not have to read the paper of course. This comment is for you as a reviewer of my work.

I understand your request for documenting how our implementation works and its relation to the official semantics. I already opened an issue #65 for that purpose. That said, this request is out of the scope of this PR. I am happy to have an informal discussion offline anytime! 

The bug you report has nothing to do with this bug fix since it has only one layer of tuples. This is related to how I extended monotonic heap from boxes to vectors. Anyway, thanks for reporting it, I will look into it.
